File: sdk/python/tests/e2e_hardening.py
"""Hardening: races, generated invariants, and malformed-input fuzzing."""

from __future__ import annotations

from concurrent.futures import ThreadPoolExecutor
from typing import Any, Dict

from harness import Server, test, unique_name, wait_until

from bunqueue import Queue, SerializationError


def generated_payloads(seed: int, count: int) -> list[Dict[str, Any]]:
    state = seed & 0xFFFFFFFF
    payloads = []
    for index in range(count):
        state = (state * 1_664_525 + 1_013_904_223) & 0xFFFFFFFF
        payloads.append(
            {
                "index": index,
                "signed": state % 2_000_001 - 1_000_000,
                "flag": bool(state & 1),
                "text": f"case-{state:x}-🧪",
                "nullable": None if index % 3 == 0 else f"value-{index}",
                "nested": [state % 97, {"checksum": (state ^ index) & 0xFFFFFFFF}],
            }
        )
    return payloads


@test
def hardening_concurrent_custom_id_retries_are_idempotent(server: Server) -> None:
    name = unique_name("idempotency-race")

    def add(attempt: int) -> str:
        with Queue(name, port=server.port) as queue:
            return queue.add(
                "charge", {"attempt": attempt}, job_id="same-operation-id"
            ).id

    with ThreadPoolExecutor(max_workers=24) as pool:
        ids = list(pool.map(add, range(24)))

    with Queue(name, port=server.port) as queue:
        try:
            assert len(set(ids)) == 1, f"concurrent retries returned {set(ids)}"
            assert queue.count() == 1, "concurrent retries enqueued a duplicate"
        finally:
            queue.obliterate()


@test
def hardening_simultaneous_dequeues_lease_exactly_once(server: Server) -> None:
    name = unique_name("double-dequeue")
    with Queue(name, port=server.port) as producer:
        expected = producer.add("only-once", {"value": 1})
        contenders = [Queue(name, port=server.port) for _ in range(12)]

        def pull(index: int) -> Dict[str, Any]:
            return contenders[index].connection.call(
                {
                    "cmd": "PULL",
                    "queue": name,
                    "owner": f"contender-{index}",
                    "timeout": 250,
                }
            )

        try:
            with ThreadPoolExecutor(max_workers=12) as pool:
                responses = list(pool.map(pull, range(12)))
            leased = [response["job"] for response in responses if response.get("job")]
            assert len(leased) == 1, f"expected one lease, got {len(leased)}"
            assert str(leased[0]["id"]) == expected.id
        finally:
            for contender in contenders:
                contender.close()
            producer.obliterate()


@test
def hardening_generated_payloads_preserve_invariants(server: Server) -> None:
    payloads = generated_payloads(0xBADC0DE, 64)
    with Queue(unique_name("generated"), port=server.port) as queue:
        try:
            ids = queue.add_bulk(
                [
                    {"name": f"generated-{index % 7}", "data": payload}
                    for index, payload in enumerate(payloads)
                ]
            )
            assert len(ids) == len(payloads)
            assert wait_until(lambda: queue.count() == len(payloads))
            for index, job_id in enumerate(ids):
                fetched = queue.get_job(job_id)
                assert fetched is not None
                data = dict(fetched.data)
                assert data.pop("name") == f"generated-{index % 7}"
                assert data == payloads[index], f"payload {index} changed: {data}"
        finally:
            queue.obliterate()


@test
def hardening_mutation_fuzz_is_typed_and_connection_recovers(server: Server) -> None:
    invalid: list[Any] = []
    for depth in range(1, 13):
        value: Any = 10 ** (1_000 + depth)
        for _level in range(depth):
            value = {"nested": [value]}
        invalid.append(value)
    cycle: Dict[str, Any] = {}
    cycle["self"] = cycle
    invalid.extend([cycle, {1: "non-string key"}])

    with Queue(unique_name("mutation-fuzz"), port=server.port) as queue:
        for payload in invalid:
            try:
                queue.connection.call({"cmd": "Ping", "payload": payload})
            except SerializationError:
                pass
            else:
                raise AssertionError("malformed mutation was accepted")
        assert not queue.connection._pending
        assert queue.ping() is True

File: sdk/typescript/tests/e2e-hardening.ts
/** Hardening: races, generated invariants, malformed-input fuzz corpus. */

import { Queue, SerializationError } from '../dist/index.js';
import { assert, assertEq, getPort, namedQueue, qname, test, waitFor } from './harness.ts';

function generator(seed: number): () => number {
  let state = seed >>> 0;
  return () => {
    state = (Math.imul(state, 1_664_525) + 1_013_904_223) >>> 0;
    return state;
  };
}

test('hardening/race: concurrent retries of one custom id enqueue exactly once', async () => {
  const name = qname('idempotency-race');
  const queues = Array.from(
    { length: 24 },
    () => new Queue<{ attempt: number }>(name, { host: '127.0.0.1', port: getPort() })
  );
  try {
    const jobs = await Promise.all(
      queues.map((queue, attempt) =>
        queue.add('charge', { attempt }, { jobId: 'same-operation-id' })
      )
    );
    assertEq(new Set(jobs.map((job) => job.id)).size, 1, 'all retries return one job id');
    assertEq(await queues[0].count(), 1, 'the broker contains exactly one job');
  } finally {
    await queues[0].obliterate();
    queues.forEach((queue) => {
      queue.close();
    });
  }
});

test('hardening/race: simultaneous dequeues lease a single job to one owner', async () => {
  const name = qname('double-dequeue');
  const producer = namedQueue(name);
  const consumers = Array.from(
    { length: 12 },
    () => new Queue(name, { host: '127.0.0.1', port: getPort() })
  );
  try {
    const expected = await producer.add('only-once', { value: 1 });
    const responses = await Promise.all(
      consumers.map((queue, index) =>
        queue.call({
          cmd: 'PULL',
          queue: name,
          owner: `contender-${index}`,
          timeout: 250,
        })
      )
    );
    const leased = responses
      .map((response) => response.job as { id?: string } | null)
      .filter((job): job is { id?: string } => job !== null);
    assertEq(leased.length, 1, 'only one contender receives a lease');
    assertEq(String(leased[0].id), expected.id, 'the unique lease belongs to the queued job');
  } finally {
    await producer.obliterate();
    producer.close();
    consumers.forEach((queue) => {
      queue.close();
    });
  }
});

test('hardening/property: generated portable payloads preserve all invariants', async () => {
  const queue = namedQueue<Record<string, unknown>>(qname('generated-payloads'));
  const next = generator(0xbadc0de);
  const payloads = Array.from({ length: 64 }, (_, index) => {
    const sample = next();
    return {
      index,
      signed: (sample % 2_000_001) - 1_000_000,
      flag: (sample & 1) === 1,
      text: `case-${sample.toString(36)}-🧪`,
      nullable: index % 3 === 0 ? null : `value-${index}`,
      nested: [sample % 97, { checksum: (sample ^ index) >>> 0 }],
    };
  });
  try {
    const jobs = await queue.addBulk(
      payloads.map((data, index) => ({ name: `generated-${index % 7}`, data }))
    );
    assertEq(jobs.length, payloads.length, 'every generated input is accepted');
    await waitFor(async () => (await queue.count()) === payloads.length);
    for (let index = 0; index < jobs.length; index++) {
      const fetched = await queue.getJob(jobs[index].id);
      assert(fetched !== null, `generated job ${index} remains queryable`);
      assertEq(fetched.name, `generated-${index % 7}`, `generated job ${index} keeps its name`);
      assertEq(
        JSON.stringify(fetched.data),
        JSON.stringify(payloads[index]),
        `generated payload ${index} round-trips exactly`
      );
    }
  } finally {
    await queue.obliterate();
    queue.close();
  }
});

test('hardening/fuzz: malformed mutations fail typed and leave the socket usable', async () => {
  const queue = namedQueue(qname('mutation-fuzz'));
  const invalid: unknown[] = [];
  for (let depth = 1; depth <= 12; depth++) {
    let value: unknown = 2n ** BigInt(32 + depth);
    for (let level = 0; level < depth; level++) value = { nested: [value] };
    invalid.push(value);
  }
  const cycle: Record<string, unknown> = {};
  cycle.self = cycle;
  invalid.push(cycle, new Map([[1, 'non-string key']]));
  try {
    for (const payload of invalid) {
      let error: unknown;
      try {
        await queue.call({ cmd: 'Ping', payload });
      } catch (caught) {
        error = caught;
      }
      assert(error instanceof SerializationError, 'malformed mutation returns SerializationError');
    }
    assert(await queue.ping(), 'the connection remains usable after the entire fuzz corpus');
  } finally {
    queue.close();
  }
});
